README.md
<!-- README.md is generated from README.Rmd. Please edit that file -->

# glmmfields <img src="inst/logo.png" align="right" />

[![R build
status](https://github.com/seananderson/glmmfields/workflows/R-CMD-check/badge.svg)](https://github.com/seananderson/glmmfields/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/glmmfields)](https://cran.r-project.org/package=glmmfields)
<!-- [![codecov](https://codecov.io/github/seananderson/glmmfields/branch/master/graphs/badge.svg)](https://codecov.io/github/seananderson/glmmfields) -->

The glmmfields R package implements Bayesian spatiotemporal models that
allow for extreme spatial deviations through time. It uses a predictive
process approach with random fields implemented through a multivariate-t
distribution instead of a multivariate normal. The models are fit with
[Stan](http://mc-stan.org/).

We published a paper describing the model and package in *Ecology*:

Anderson, S. C., Ward, E. J. 2019. Black swans in space: modelling
spatiotemporal processes with extremes. 100(1):e02403.
<https://doi.org/10.1002/ecy.2403>

You can install the [CRAN
version](https://cran.r-project.org/package=glmmfields) of the package
with:

``` r
install.packages("glmmfields")
```

If you have a C++ compiler installed, you can install the development
version of the package with:

``` r
# install.packages("remotes")
remotes::install_github("seananderson/glmmfields", build_vignettes = TRUE)
```

glmmfields can also fit spatial GLMs with Stan. See the vignette:

``` r
vignette("spatial-glms", package = "glmmfields")
```

## An example spatiotemporal model

``` r
library(glmmfields)
#> Loading required package: Rcpp
library(ggplot2)
```

Simulate data:

``` r
set.seed(42)
s <- sim_glmmfields(
  df = 2.8, n_draws = 12, n_knots = 12, gp_theta = 2.5,
  gp_sigma = 0.2, sd_obs = 0.1
)
head(s$dat)
#>   time pt           y      lon      lat station_id
#> 1    1  1  0.02818963 9.148060 6.262453          1
#> 2    1  2 -0.21924739 9.370754 2.171577          2
#> 3    1  3 -0.34719485 2.861395 2.165673          3
#> 4    1  4 -0.15785483 8.304476 3.889450          4
#> 5    1  5 -0.04703617 6.417455 9.424557          5
#> 6    1  6 -0.23904924 5.190959 9.626080          6
```

``` r
print(s$plot)
```

![](README-figs/plot-sim-1.png)<!-- -->

Fit the model:

``` r
options(mc.cores = parallel::detectCores()) # for parallel processing
m <- glmmfields(y ~ 0,
  data = s$dat, time = "time",
  lat = "lat", lon = "lon",
  nknots = 12, estimate_df = TRUE, iter = 800, seed = 1
)
```

``` r
print(m)
#> Inference for Stan model: glmmfields.
#> 4 chains, each with iter=800; warmup=400; thin=1; 
#> post-warmup draws per chain=400, total post-warmup draws=1600.
#> 
#>             mean se_mean   sd    2.5%     25%     50%     75%   97.5% n_eff Rhat
#> df[1]       3.72    0.04 1.47    2.08    2.67    3.37    4.28    7.48  1331    1
#> gp_sigma    0.30    0.00 0.04    0.22    0.27    0.30    0.32    0.39   525    1
#> gp_theta    2.58    0.00 0.07    2.46    2.54    2.58    2.63    2.71  1434    1
#> sigma[1]    0.10    0.00 0.00    0.09    0.10    0.10    0.10    0.10  2207    1
#> lp__     2291.28    0.42 9.59 2270.34 2285.12 2291.59 2297.73 2308.74   521    1
#> 
#> Samples were drawn using NUTS(diag_e) at Mon Feb 13 12:45:42 2023.
#> For each parameter, n_eff is a crude measure of effective sample size,
#> and Rhat is the potential scale reduction factor on split chains (at 
#> convergence, Rhat=1).
```

Plot:

``` r
plot(m, type = "prediction") + scale_color_gradient2()
```

![](README-figs/plot-predictions-1.png)<!-- -->

``` r
plot(m, type = "spatial-residual")
```

![](README-figs/plot-predictions-2.png)<!-- -->

Predictions:

``` r
# link scale:
p <- predict(m)
head(p)
#> # A tibble: 6 × 3
#>   estimate conf_low conf_high
#>      <dbl>    <dbl>     <dbl>
#> 1  -0.0283  -0.0868    0.0273
#> 2  -0.291   -0.365    -0.220 
#> 3  -0.397   -0.448    -0.346 
#> 4  -0.196   -0.266    -0.123 
#> 5  -0.0370  -0.110     0.0360
#> 6  -0.214   -0.294    -0.140

# posterior predictive intervals on new observations (include observation error):
p <- predictive_interval(m)
head(p)
#> # A tibble: 6 × 3
#>   estimate conf_low conf_high
#>      <dbl>    <dbl>     <dbl>
#> 1  -0.0283   -0.236   0.181  
#> 2  -0.291    -0.507  -0.0904 
#> 3  -0.397    -0.596  -0.206  
#> 4  -0.196    -0.392   0.00154
#> 5  -0.0370   -0.239   0.172  
#> 6  -0.214    -0.423  -0.00659
```

Use the `tidy` method to extract parameter estimates as a data frame:

``` r
x <- tidy(m, conf.int = TRUE)
head(x)
#> # A tibble: 6 × 5
#>   term                     estimate std.error conf.low conf.high
#>   <chr>                       <dbl>     <dbl>    <dbl>     <dbl>
#> 1 df[1]                      3.37     1.47      2.08      7.48  
#> 2 gp_sigma                   0.295    0.0432    0.216     0.388 
#> 3 gp_theta                   2.58     0.0662    2.46      2.71  
#> 4 sigma[1]                   0.0979   0.00214   0.0939    0.102 
#> 5 spatialEffectsKnots[1,1]  -0.110    0.0341   -0.175    -0.0442
#> 6 spatialEffectsKnots[2,1]  -0.230    0.0386   -0.305    -0.155
```

Make predictions on a fine-scale spatial grid:

``` r
pred_grid <- expand.grid(
  lat = seq(min(s$dat$lat), max(s$dat$lat), length.out = 25),
  lon = seq(min(s$dat$lon), max(s$dat$lon), length.out = 25),
  time = unique(s$dat$time)
)

pred_grid$prediction <- predict(m,
  newdata = pred_grid, type = "response", iter = 100, estimate_method = "median"
)$estimate

ggplot(pred_grid, aes(lon, lat, fill = prediction)) +
  facet_wrap(~time) +
  geom_raster() +
  scale_fill_gradient2()
```

![](README-figs/grid-predictions-1.png)<!-- -->

# References

Anderson, S. C., Ward, E. J. 2019. Black swans in space: modelling
spatiotemporal processes with extremes. 100(1):e02403.
<https://doi.org/10.1002/ecy.2403>

Latimer, A. M., S. Banerjee, H. Sang Jr, E. S. Mosher, and J. A.
Silander Jr. 2009. Hierarchical models facilitate spatial analysis of
large data sets: a case study on invasive plant species in the
northeastern United States. Ecology Letters 12:144–154.

Shelton, A. O., J. T. Thorson, E. J. Ward, and B. E. Feist. 2014.
Spatial semiparametric models improve estimates of species abundance and
distribution. Canadian Journal of Fisheries and Aquatic Sciences
71:1655–1666.

### NOAA Disclaimer

This repository is a scientific product and is not official
communication of the National Oceanic and Atmospheric Administration, or
the United States Department of Commerce. All NOAA GitHub project code
is provided on an ‘as is’ basis and the user assumes responsibility for
its use. Any claims against the Department of Commerce or Department of
Commerce bureaus stemming from the use of this GitHub project will be
governed by all applicable Federal law. Any reference to specific
commercial products, processes, or services by service mark, trademark,
manufacturer, or otherwise, does not constitute or imply their
endorsement, recommendation or favoring by the Department of Commerce.
The Department of Commerce seal and logo, or the seal and logo of a DOC
bureau, shall not be used in any manner to imply endorsement of any
commercial product or activity by DOC or the United States Government.


NEWS.md
# glmmfields 0.1.8

* Update for rstan 2.26 (#16)

# glmmfields 0.1.7

* Rebuild with newest rstantools to avoid CRAN NOTE about C++ version

# glmmfields 0.1.6

* Add back RcppParallel as an import

# glmmfields 0.1.5

* Fully delegate installation to rstantools (#15)

* Other minor fixes to pass R CMD check on R devel

* Add offset functionality

# glmmfields 0.1.4

* Make compatible with updated broom.mixed

# glmmfields 0.1.3

* Make compatible with C++14.

# glmmfields 0.1.2

* Make compatible with R 3.6.0 staged installation and latest rstantools.

# glmmfields 0.1.1

* Changed how Stan finds directories / files

# glmmfields 0.1.0.9000

* Add support for random walk year effects with covariates. There are a few
  specific cases where covariates or the random year effect are not estimable.
  Examples are:

    1. estimating an intercept and AR year effects (intercept confounded with
       1st year effect)
    2. estimating AR year effects and estimating 'phi' — the AR parameter in
       spatial field (also confounded)

* Import S3 methods from rstantools instead of rstanarm (#5)

* Adjust calculation of year index values to better allow for missing time slices

# glmmfields 0.1.0

* Initial submission to CRAN.


R/predict.R
#' Predict from a glmmfields model
#'
#' These functions extract posterior draws or credible intervals. The helper
#' functions are named to match those in the \pkg{rstanarm} package and call the
#' function `predict()` with appropriate argument values.
#'
#' @param object An object returned by [glmmfields()].
#' @param newdata Optionally, a data frame to predict on
#' @param interval Type of interval calculation. Same as for
#'   [stats::predict.lm()].
#' @param estimate_method Method for computing point estimate ("mean" or
#'   "median")
#' @param conf_level Probability level for the credible intervals.
#' @param type Whether the predictions are returned on "link" scale or
#'   "response" scale (Same as for [stats::predict.glm()]).
#' @param return_mcmc Logical. Should the full MCMC draws be returned for the
#'   predictions?
#' @param offset Optional offset vector to be used in prediction.
#' @param iter Number of MCMC iterations to draw. Defaults to all.
#' @param ... Ignored currently
#'
#' @importFrom stats median quantile rgamma rnbinom
#' @importFrom assertthat assert_that
#'
#' @examples
#' \donttest{
#' library(ggplot2)
#'
#' # simulate:
#' set.seed(1)
#' s <- sim_glmmfields(
#'   n_draws = 12, n_knots = 12, gp_theta = 2.5,
#'   gp_sigma = 0.2, sd_obs = 0.1
#' )
#'
#' # fit:
#' # options(mc.cores = parallel::detectCores()) # for parallel processing
#' m <- glmmfields(y ~ 0,
#'   data = s$dat, time = "time",
#'   lat = "lat", lon = "lon",
#'   nknots = 12, iter = 800, chains = 1
#' )
#'
#' # Predictions:
#' # Link scale credible intervals:
#' p <- predict(m, type = "link", interval = "confidence")
#' head(p)
#'
#' # Prediction intervals on new observations (include observation error):
#' p <- predictive_interval(m)
#' head(p)
#'
#' # Posterior prediction draws:
#' p <- posterior_predict(m, iter = 100)
#' dim(p) # rows are iterations and columns are data elements
#'
#' # Draws from the linear predictor (not in link space):
#' p <- posterior_linpred(m, iter = 100)
#' dim(p) # rows are iterations and columns are data elements
#'
#' # Use the `tidy` method to extract parameter estimates as a data frame:
#' head(tidy(m, conf.int = TRUE, conf.method = "HPDinterval"))
#'
#' # Make predictions on a fine-scale spatial grid:
#' pred_grid <- expand.grid(
#'   lat = seq(min(s$dat$lat), max(s$dat$lat), length.out = 25),
#'   lon = seq(min(s$dat$lon), max(s$dat$lon), length.out = 25),
#'   time = unique(s$dat$time)
#' )
#' pred_grid$prediction <- predict(m,
#'   newdata = pred_grid, type = "response", iter = 100,
#'   estimate_method = "median", offset = rep(0, nrow(pred_grid))
#' )$estimate
#'
#' ggplot(pred_grid, aes(lon, lat, fill = prediction)) +
#'   facet_wrap(~time) +
#'   geom_raster() +
#'   scale_fill_gradient2()
#' }
#' @name predict
NULL

#' @name predictive_interval
#' @rdname predict
#' @export
#' @importFrom rstantools predictive_interval
NULL

#' @name posterior_linpred
#' @rdname predict
#' @export
#' @importFrom rstantools posterior_linpred
NULL

#' @name posterior_predict
#' @rdname predict
#' @export
#' @importFrom rstantools posterior_predict
NULL

#' @rdname predict
#' @export
predictive_interval.glmmfields <- function(object, ...) {
  predict.glmmfields(object, interval = "prediction", type = "response",
    return_mcmc = FALSE, ...)
}

#' @rdname predict
#' @export
posterior_linpred.glmmfields <- function(object, ...) {
  predict.glmmfields(object, interval = "confidence", type = "response",
    return_mcmc = TRUE, ...)
}

#' @rdname predict
#' @export
posterior_predict.glmmfields <- function(object, ...) {
  predict.glmmfields(object, interval = "prediction", type = "response",
    return_mcmc = TRUE, ...)
}

#' @importFrom stats predict
#' @rdname predict
#' @export
predict.glmmfields <- function(object, newdata = NULL,
                               estimate_method = c("median", "mean"),
                               conf_level = 0.95,
                               interval = c("confidence", "prediction"),
                               type = c("link", "response"),
                               return_mcmc = FALSE,
                               offset = NULL,
                               iter = "all", ...) {
  estimate_method <- match.arg(estimate_method)
  interval <- match.arg(interval)
  type <- match.arg(type)

  assert_that(is.numeric(conf_level))
  assert_that(identical(length(conf_level), 1L))
  assert_that(conf_level > 0 && conf_level < 1)
  assert_that(identical(class(object), "glmmfields"))

  if (interval == "prediction" && type != "response") {
    stop("type must be 'response' if interval is 'prediction")
  }

  obs_model <- object$obs_model

  if (is.null(newdata)) offset <- object$offset
  if (is.null(offset) && !is.null(newdata)) offset <- rep(0, nrow(newdata))
  if (!is.null(newdata) && !is.null(object$offset)) {
    if (is.null(offset)) stop("Missing `offset` argument.", call. = FALSE)
  }

  # newdata is df with time, y, lon, lat
  # if null, defaults to data used to fit model
  if (is.null(newdata)) newdata <- object$data

  response <- all.vars(nlme::getResponseFormula(object$formula))
  newdata[[response]] <- 1.0

  newdata <- tibble::as_tibble(newdata)
  # create model.matrix() as in fitting function, only with newdata
  X <- model.matrix(object$formula,
    model.frame(object$formula, newdata, na.action = na.omit))

  .missing <- colnames(object$X)[!colnames(object$X) %in% colnames(X)]
  if (length(.missing) > 0L)
    stop(paste(.missing, collapse = ", "), " are missing in `newdata`.")

  if (nrow(X) < nrow(newdata))
    stop("Some predictors in newdata had NA values.")

  time <- object$time
  knots <- object$knots

  dist_knots <- as.matrix(dist(knots))

  # Calculate distance from knots to grid
  dist_all <- as.matrix(stats::dist(rbind(
    newdata[, c(object$lon, object$lat)],
    knots
  )))
  n_locs <- nrow(newdata)

  # this is the transpose of the lower left corner
  dist_knots21 <- t(
    dist_all[-seq_len(n_locs), -seq(n_locs + 1, ncol(dist_all))]
  )

  # extract mcmc pars
  pars <- rstan::extract(object$model, permuted = TRUE)
  ##
  if (iter == "all") {
    mcmc.i <- seq_len(length(pars$lp__))
  } else {
    mcmc.i <- base::sample(seq_len(length(pars$lp__)), size = iter)
  }
  mcmc_draws <- length(mcmc.i)
  pred_values <- matrix(NA, n_locs, mcmc_draws)
  for (i in seq_len(mcmc_draws)) {
    # create cov matrix at knots
    if (object$covariance == "exponential") {
      covmat <- pars$gp_sigma[mcmc.i[i]] *
        exp(-dist_knots / pars$gp_theta[mcmc.i[i]])
      covmat21 <- pars$gp_sigma[mcmc.i[i]] *
        exp(-dist_knots21 / pars$gp_theta[mcmc.i[i]])
    }
    if (object$covariance == "squared-exponential") {
      covmat <- pars$gp_sigma[mcmc.i[i]] *
        exp(-(dist_knots^2) / (2 * pars$gp_theta[mcmc.i[i]]^2))
      covmat21 <- pars$gp_sigma[mcmc.i[i]] *
        exp(-(dist_knots21^2) / (2 * pars$gp_theta[mcmc.i[i]]^2))
    }
    if (object$covariance == "matern") {
      if (object$matern_kappa == 1.5) {
        transformed_dist <- sqrt(3) * dist_knots / pars$gp_theta[mcmc.i[i]]
        covmat <-
          pars$gp_sigma[mcmc.i[i]] *
            (1 + transformed_dist) * exp(-transformed_dist)

        transformed_dist <- sqrt(3) * dist_knots21 / pars$gp_theta[mcmc.i[i]]
        covmat21 <-
          pars$gp_sigma[mcmc.i[i]] *
            (1 + transformed_dist) * exp(-transformed_dist)
      }
      if (object$matern_kappa == 2.5) {
        transformed_dist <- sqrt(5) * dist_knots / pars$gp_theta[mcmc.i[i]]
        covmat <-
          pars$gp_sigma[mcmc.i[i]] *
            (1 + transformed_dist + (transformed_dist^2) / 3) * exp(-transformed_dist)

        transformed_dist <- sqrt(5) * dist_knots21 / pars$gp_theta[mcmc.i[i]]
        covmat21 <-
          pars$gp_sigma[mcmc.i[i]] *
            (1 + transformed_dist + (transformed_dist^2) / 3) * exp(-transformed_dist)
      }
    }
    # these are projected spatial effects, dim = new data points x time
    spat_eff_knots_i <- pars$spatialEffectsKnots[mcmc.i[i], , ]
    if (is.matrix(spat_eff_knots_i)) {
      spat_eff_knots_i <- t(spat_eff_knots_i)
    } else { # these are for one time slice and are a vector
      spat_eff_knots_i <- t(t(spat_eff_knots_i))
    }
    spat_effects <- covmat21 %*% solve(covmat) %*% spat_eff_knots_i

    rows <- seq_len(n_locs)

    if (identical(time, "null_time_")) {
      newdata$null_time_ <- 1
      time <- "null_time_"
    }
    cols <- as.numeric(as.factor(newdata[, time][[1]]))
    # check this for > 1 year. B will also have to be modified
    if (!object$year_re) {
      if (!object$fixed_intercept) {
        pred_values[, i] <- X %*% matrix(pars$B[mcmc.i[i], ], ncol = 1) +
          spat_effects[cbind(rows, cols)]
      } else {
        pred_values[, i] <- spat_effects[cbind(rows, cols)]
      }
    } else {
      pred_values[, i] <- spat_effects[cbind(rows, cols)] +
        pars$yearEffects[mcmc.i[i], ][cols]
    }
  }

  mcmc_draws <- ncol(pred_values)

  # if type == link, don't include observation/data model.

  # If predictions other than on link scale, use observation model and link to
  # generate (1) confidence intervals on mean or (2) prediction intervals
  # including obs error

  pred_values <- pred_values + offset

  if (type == "response") {
    # gamma or NB2 or poisson:
    if (obs_model %in% c(0, 2, 5, 6)) pred_values <- exp(pred_values)

    if (obs_model == 1) {
      # normal, sigma is returned
      pp <- t(apply(pred_values, 1, function(x)
        stats::rnorm(mcmc_draws, mean = x, sd = pars$sigma[, 1])))
    }

    # binomial (plogis = inverse logit):
    if (obs_model == 4) pred_values <- stats::plogis(pred_values)

    if (obs_model == 0) {
      # gamma, CV is returned; gammaA = 1/(CV*CV)
      pp <- t(apply(pred_values, 1, function(x)
        stats::rgamma(mcmc_draws,
          shape = 1 / (pars$CV[, 1]^2),
          rate = 1 / (pars$CV[, 1]^2) / x
        )))
    }
    if (obs_model == 2) {
      # negative binomial, phi returned
      pp <- t(apply(pred_values, 1, function(x)
        stats::rnbinom(mcmc_draws, mu = x, size = pars$nb2_phi[, 1])))
    }
    if (obs_model == 4) {
      # binomial
      pp <- t(apply(pred_values, 1, function(x)
        stats::rbinom(mcmc_draws, size = 1, prob = x)))
    }
    if (obs_model == 5) {
      # poisson
      pp <- t(apply(pred_values, 1, function(x)
        stats::rpois(mcmc_draws, lambda = x)))
    }
    if (obs_model == 6) {
      # lognormal, sigma is returned
      pp <- t(apply(pred_values, 1, function(x)
        stats::rlnorm(mcmc_draws,
          meanlog = log(x),
          sdlog = pars$sigma[, 1]
        )))
    }
  }

  est_method <- switch(estimate_method[[1]], median = median, mean = mean)
  out <- data.frame(estimate = apply(pred_values, 1, est_method))

  if (interval == "confidence") {
    out$conf_low <- apply(pred_values, 1, quantile,
      probs = (1 - conf_level) / 2
    )
    out$conf_high <- apply(pred_values, 1, quantile,
      probs = 1 - (1 - conf_level) / 2
    )
  }
  if (interval == "prediction" && type == "response") {
    out$conf_low <- apply(pp, 1, quantile, probs = (1 - conf_level) / 2)
    out$conf_high <- apply(pp, 1, quantile, probs = 1 - (1 - conf_level) / 2)
  }

  out <- tibble::as_tibble(out)
  if (return_mcmc) out <- t(pred_values) # to match rstanarm generic methods
  out
}
